sql: notices for NotVisible Indexes

wenyihu6 opened the pull request:
Optimizer now supports creating invisible indexes after this
[PR](https://github.com/cockroachdb/cockroach/pull/85794). An important use case
for not visible indexes is to test the behaviour of dropping an index by marking
the index invisible. However, there are certain cases where users cannot expect
dropping an index to behave exactly the same as marking an index invisible. More
specifically, NotVisible indexes may still be used to police unique or foreign
key constraint check behind the scene. In those cases, dropping the index might
behave different from marking the index invisible. Prior to this commit, users
do not know about this without reading the documentation. This commit adds some
user-friendly notices when users are dropping or changing a not visible index
that might be helpful for constraint check.

There are two cases where we are giving this notice: 1. if this index is unique.
2. if this index is on child table and may help with FK check.

More details on how this decision was made in
docs/RFCS/20220628_invisible_index.md.

Assists: https://github.com/cockroachdb/cockroach/issues/72576

See also: https://github.com/cockroachdb/cockroach/pull/85794

Release justification: low risk to the existing functionality; this commit just
adds notices.

Release note: none

wenyihu6 committed:
sql: notices for NotVisible Indexes

Optimizer now supports creating invisible indexes after this
[PR](https://github.com/cockroachdb/cockroach/pull/85794). An important use case
for not visible indexes is to test the behaviour of dropping an index by marking
the index invisible. However, there are certain cases where users cannot expect
dropping an index to behave exactly the same as marking an index invisible. More
specifically, NotVisible indexes may still be used to police unique or foreign
key constraint check behind the scene. In those cases, dropping the index might
behave different from marking the index invisible. Prior to this commit, users
do not know about this without reading the documentation. This commit adds some
user-friendly notices when users are dropping or changing a not visible index
that might be helpful for constraint check.

There are two cases where we are giving this notice: 1. if this index is unique.
2. if this index is on child table and may help with FK check.

More details on how this decision was made in
docs/RFCS/20220628_invisible_index.md.

Assists: https://github.com/cockroachdb/cockroach/issues/72576

See also: https://github.com/cockroachdb/cockroach/pull/85794

Release justification: low risk to the existing functionality; this commit just
adds notices.

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85354)
<!-- Reviewable:end -->

wenyihu6 commented:
The first commit comes from https://github.com/cockroachdb/cockroach/pull/84912. The second commit comes from https://github.com/cockroachdb/cockroach/pull/85239.

michae2 reviewed (dismissed):
Nice work! I really appreciate all the testcases!

The notices are a nice touch, but I think it's a little too noisy to return them every time there's an invisible index and a FK reference in the same table. I'd like them to be more specific, and only fire when the invisible index could be used by the FK reference. I sketched out some code that should do that below.

> Reviewed 1 of 10 files at r7, 6 of 9 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-:-N8jQAkhExjedb7wMGJH:b56oxdk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rhu713, and @wenyihu6)

___
*[`pkg/sql/create_index.go` line 745 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8j9j_p-nc8F8R7lIwG:-N8j9j_p-nc8F8R7lIwH:b-vjbsot) ([raw file](https://github.com/cockroachdb/cockroach/blob/6082e6c2e75f04b9b049c5c3205cea10c579e55e/pkg/sql/create_index.go#L745)):*
> ```Go
> 		// the scene. In this case, dropping the index may behave differently than
> 		// marking the index invisible.
> 		if n.n.Unique || len(n.tableDesc.OutboundFKs) != 0 {
> ```

nit: I suggest we change `len(n.tableDesc.OutboundFKs) != 0` to be a little more specific, something like:

```
var isValidOutboundFKIndex bool
for _, fk := range n.tableDesc.OutboundFKs {
  if (fk.OnDelete != ForeignKeyAction_NO_ACTION || fk.OnUpdate != ForeignKeyAction_NO_ACTION) && indexDesc.IsValidOriginIndex(fk.OriginColumnIDs) {
    isValidOutboundFKIndex = true
    break
  }
}
```
___
*[`pkg/sql/create_table.go` line 1796 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8jM88f3tdItX39W3AS:-N8jM88f3tdItX39W3AT:b-5p77d4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6082e6c2e75f04b9b049c5c3205cea10c579e55e/pkg/sql/create_table.go#L1796)):*
> ```Go
> 	// This flag is created to avoid sending multiple notices if user is adding
> 	// multiple FK constraints or multiple unique invisible indexes.
> 	sentNotVisibleMsgAlready := new(bool)
> ```

Instead of placing duplicate checks and notices in FK resolution and unique index creation, let's hoist the check and notice up a level and run it just once for the entire table after everything is ready. I suggest we put it somewhere near the end of `NewTableDesc`, probably just after the call to `tabledesc.ValidateOnUpdate`.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/create_index.go` line 745 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8j9j_p-nc8F8R7lIwG:-N8jemYO6xyMEAJLuw8o:b9qx4ud) ([raw file](https://github.com/cockroachdb/cockroach/blob/6082e6c2e75f04b9b049c5c3205cea10c579e55e/pkg/sql/create_index.go#L745)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: I suggest we change `len(n.tableDesc.OutboundFKs) != 0` to be a little more specific, something like:

```
var isValidOutboundFKIndex bool
for _, fk := range n.tableDesc.OutboundFKs {
  if (fk.OnDelete != ForeignKeyAction_NO_ACTION || fk.OnUpdate != ForeignKeyAction_NO_ACTION) && indexDesc.IsValidOriginIndex(fk.OriginColumnIDs) {
    isValidOutboundFKIndex = true
    break
  }
}
```
</blockquote></details>

Ahh i see. Thanks for looking into this! I wanted to do something like this but didn't know we have this nice trick IsValidOriginIndex to check if the index will be actually used for fk check.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/create_table.go` line 1796 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8jM88f3tdItX39W3AS:-N8jfNjm22tspi5sArci:b-m55fbu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6082e6c2e75f04b9b049c5c3205cea10c579e55e/pkg/sql/create_table.go#L1796)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Instead of placing duplicate checks and notices in FK resolution and unique index creation, let's hoist the check and notice up a level and run it just once for the entire table after everything is ready. I suggest we put it somewhere near the end of `NewTableDesc`, probably just after the call to `tabledesc.ValidateOnUpdate`.
</blockquote></details>

Sounds good! I will work on them now.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/create_index.go` line 745 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8j9j_p-nc8F8R7lIwG:-N8vZdO34LTVnQhZGt3j:b-hx6wns) ([raw file](https://github.com/cockroachdb/cockroach/blob/6082e6c2e75f04b9b049c5c3205cea10c579e55e/pkg/sql/create_index.go#L745)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Ahh i see. Thanks for looking into this! I wanted to do something like this but didn't know we have this nice trick IsValidOriginIndex to check if the index will be actually used for fk check.
</blockquote></details>

Thank you! I moved the check to `pkg/sql/catalog/tabledesc/validate.go`. Every time after table has been fully updated in `create_table.go`, `alter_table.go`, and `create_index.go`, I pass in the new table and call this check `ValidateNotVisibleIndexWithinTable` to know if we should give a notice. This way, we don't need to handle deduplication of notices. 

A thing I would like to double check is how we should predict if an index on a child table may become helpful for FK check. I have discussed this with you on slack as well; I just want to document the discussion on this PR. As of now, I am checking if the index's `KeyColumnIDs` has intersection with the FK constraint's `originColIDs`.

1. On the first thought, we could use `IsValidOriginIndex `. Based on its comment (IsValidOriginIndex returns whether the index can serve as an origin index for a foreign key constraint with the provided set of originColIDs), it should be doing what we want here. But based on the code, it is checking if `originColIDs` is the prefix of the index's `KeyColumnIDs`. I'm not sure if that's what we want. Consider the following case:  a, b might not be a, c, b's prefix, but it's still useful for constraint check that reference column a, b.
![Screen Shot 2022-08-08 at 12.28.05 AM.png](https://files.reviewableusercontent.io/images/reviewable/56973754/eWYU3zRXJhd2Bm9b5yXeB932/Screen%20Shot%202022-08-08%20at%2012.28.05%20AM.png)
![Screen Shot 2022-08-07 at 11.53.55 PM.png](https://files.reviewableusercontent.io/images/reviewable/56973754/DYjcOjkhOOHVqjzcUeZhhC1J/Screen%20Shot%202022-08-07%20at%2011.53.55%20PM.png

Now the question turns to whether `originColIDs` is an ordered list. This is because:`IsValidOriginIndex` is checking if `originColIDs` is the prefix of the index's `KeyColumnIDs`. Based on the comment for `KeyColumnIDs`,`KeyColumnIDs` is an ordered list of column ids. If `originColIDs` is also an ordered list, then this function is not just checking for `KeyColumnIDs`'s prefix but also checking if `KeyColumnIDs` will cover `originColIDs` (`originColIDs`will be a subset of `KeyColumnIDs`). Then the example in the screenshot would be fine. If `originColIDs` is not an ordered list but just in the same order that the user specified, this might cause problems mentioned in the screenshot.

I'm not exactly sure how `IsValidOriginIndex ` and `HasPrefix` are being used right now; I took a quick look at it, but I need more time to understand better.

2. Should we check if `KeyColumnIDs` will cover `originColIDs` (if `originColIDs` is a subset of `KeyColumnIDs`) or if their intersection is not empty? Consider the following example: idx covers a (has intersection as the constraint column). But it's not helpful here since the constraint reference comes from column a and b. I injected a lot of stats here, but it's still favouring primary index over idx. I'm not sure if this is supposed to happen. When the table gets really large, will it ever favour the the secondary index and perform a look up join?


![Screen Shot 2022-08-08 at 12.19.47 AM.png](https://files.reviewableusercontent.io/images/reviewable/56973754/Zxda4T4-6IYZL1aAkpv7PT2a/Screen%20Shot%202022-08-08%20at%2012.19.47%20AM.png)

Based on the experiment with the demo here, checking if `KeyColumnIDs` will cover `originColIDs` (if `originColIDs` is a subset of `KeyColumnIDs`) seems more correct here. But I'm not sure if I am missing anything.

Sorry if I am making things really confusing here. Please let me know if any parts is confusing and I will try rephrasing.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Thank you! I moved the check to `pkg/sql/catalog/tabledesc/validate.go`. Every time after table has been fully updated in `create_table.go`, `alter_table.go`, and `create_index.go`, I pass in the new table and call this check `ValidateNotVisibleIndexWithinTable` to know if we should give a notice. This way, we don't need to handle deduplication of notices. 

A thing I would like to double check is how we should predict if an index on a child table may become helpful for FK check. I have discussed this with you on slack as well; I just want to document the discussion on this PR. As of now, I am checking if the index's `KeyColumnIDs` has intersection with the FK constraint's `originColIDs`.

1. On the first thought, we could use `IsValidOriginIndex `. Based on its comment (IsValidOriginIndex returns whether the index can serve as an origin index for a foreign key constraint with the provided set of originColIDs), it should be doing what we want here. But based on the code, it is checking if `originColIDs` is the prefix of the index's `KeyColumnIDs`. I'm not sure if that's what we want. Consider the following case:  a, b might not be a, c, b's prefix, but it's still useful for constraint check that reference column a, b.
![Screen Shot 2022-08-08 at 12.28.05 AM.png](https://files.reviewableusercontent.io/images/reviewable/56973754/eWYU3zRXJhd2Bm9b5yXeB932/Screen%20Shot%202022-08-08%20at%2012.28.05%20AM.png)


Now the question turns to whether `originColIDs` is an ordered list. This is because:`IsValidOriginIndex` is checking if `originColIDs` is the prefix of the index's `KeyColumnIDs`. Based on the comment for `KeyColumnIDs`,`KeyColumnIDs` is an ordered list of column ids. If `originColIDs` is also an ordered list, then this function is not just checking for `KeyColumnIDs`'s prefix but also checking if `KeyColumnIDs` will cover `originColIDs` (`originColIDs`will be a subset of `KeyColumnIDs`). Then the example in the screenshot would be fine. If `originColIDs` is not an ordered list but just in the same order that the user specified, this might cause problems mentioned in the screenshot.

I'm not exactly sure how `IsValidOriginIndex ` and `HasPrefix` are being used right now; I took a quick look at it, but I need more time to understand better.

2. Should we check if `KeyColumnIDs` will cover `originColIDs` (if `originColIDs` is a subset of `KeyColumnIDs`) or if their intersection is not empty? Consider the following example: idx covers a (has intersection as the constraint column). But it's not helpful here since the constraint reference comes from column a and b. I injected a lot of stats here, but it's still favouring primary index over idx. I'm not sure if this is supposed to happen. When the table gets really large, will it ever favour the the secondary index and perform a look up join?


![Screen Shot 2022-08-08 at 12.19.47 AM.png](https://files.reviewableusercontent.io/images/reviewable/56973754/Zxda4T4-6IYZL1aAkpv7PT2a/Screen%20Shot%202022-08-08%20at%2012.19.47%20AM.png)

Based on the experiment with the demo here, checking if `KeyColumnIDs` will cover `originColIDs` (if `originColIDs` is a subset of `KeyColumnIDs`) seems more correct here. But I'm not sure if I am missing anything.

Sorry if I am making things really confusing here. Please let me know if anything is confusing here and I can try rephrasing.





<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
I left a few comments on the notice-related logic that might be helpful, but we should talk about it IRL.

Great work on this! Nice to see everything coming together!

All the NOTICE-related changes would be a good candidate to break into a separate commit or PR. It's up to you whether you want to split them up now, but something to keep in mind for future PRs.

> Reviewed 1 of 9 files at r12, 5 of 56 files at r13, 5 of 6 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-:-N8zkMnxCkFs_UzaUZw-:bu6zyah)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @rhu713, and @wenyihu6)

___
*[`pkg/sql/alter_table.go` line 865 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8ze8HLBYsBdNcg2LZM:-N8ze8HLBYsBdNcg2LZN:b1sw8hl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/alter_table.go#L865)):*
> ```Go
> 	// Warn against dropping an index if there exists a NotVisible index that may
> 	// be used for constraint check behind the scene.
> 	if notVisibleIndexNotice := tabledesc.ValidateNotVisibleIndexWithinTable(n.tableDesc); notVisibleIndexNotice != nil {
> ```

I don't understand this comment and code. Don't we only want to show a notice if an invisible index is being dropped that could be helpful for FK checks and cascades?
___
*[`pkg/sql/catalog/descpb/index.go` line 73 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zce2SAHplPD_OGAqR:-N8zce2SAHplPD_OGAqS:b-ui1hxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/catalog/descpb/index.go#L73)):*
> ```Go
> // TODO(wenyihu6): check if checking HasIntersection is correct here.
> func (desc *IndexDescriptor) IsHelpfulOriginIndex(originColIDs ColumnIDs) bool {
> 	return !desc.IsPartial() && ColumnIDs(desc.KeyColumnIDs).HasIntersection(originColIDs)
> ```

I'm wondering if we'd reduce some noise from these notices by checking that a prefix of the indexes columns are included in the FK constraint. I don't think that'd miss any cases where an index would be helping FK checks and cascades. For example:

Helpful:

```
INDEX(a, b, c)
FK (a, b)
```

Helpful:

```
INDEX(a, b, c)
FK (b, a)
```

Not helpful:

```
INDEX(a, b, c)
FK (b, c)
```
___
*[`pkg/sql/catalog/descpb/structured.go` line 152 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zcOSLCqo48Vq-oNLw:-N8zcOSLCqo48Vq-oNLx:b-jdbuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/catalog/descpb/structured.go#L152)):*
> ```Go
> 
> // HasIntersection returns true if the input list has intersection with columnIDs.
> func (c ColumnIDs) HasIntersection(input ColumnIDs) bool {
> ```

nit: You can name this `Intersects` to match similar functions like `FastIntSet.Intersects`.
___
*[`pkg/sql/catalog/descpb/structured.go` line 153 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zbKLIBbyWLNA8msEN:-N8zbKLIBbyWLNA8msEO:b-asv7yk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/catalog/descpb/structured.go#L153)):*
> ```Go
> // HasIntersection returns true if the input list has intersection with columnIDs.
> func (c ColumnIDs) HasIntersection(input ColumnIDs) bool {
> 	return c.UtilFastIntSetOperation(input, util.FastIntSet.Intersects)
> ```

We're likely dealing with relatively small slices of column IDs, so it'd probably be faster to use nested loops to look for an intersection. PermutationOf requires extra book-keeping and needs to work regardless of the order of column IDs, which we don't need here. Also, we can return early as soon as we find a single matching column, whereas PermutationOf has to consider all columns in the slices.
___
*[`pkg/sql/catalog/tabledesc/index.go` line 146 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zdNNN8olmNSG8s8bD:-N8zdNNN8olmNSG8s8bE:b-hyn2vu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/catalog/tabledesc/index.go#L146)):*
> ```Go
> 
> // IsHelpfulOriginIndex returns whether the index may become a helpful index for
> // foreign key constraint check on the child table.
> ```

nit: "may become a helpful index for foreign key constraint check on the child table" => "may be a helpful index for performing foreign key checks and cascades for a foreign key with the given origin columns"
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 15 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zecZIB747kX22uyHe:-N8zecZIB747kX22uyHf:b-xy5625) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L15)):*
> ```
> #   - Whenever an invisible index is created on a child table in CREATE INDEX or CREATE TABLE
> #   - Whenever a new foreign key constraint is added to a table with invisible index
> #   - Whenever they are using force index with invisible index
> ```

nit: AFAIK the standard convention is put tests that output query plans (or output kvtraces) in these execbuilder tests, and all other types of tests, like the `SHOW ...` tests in a file like `pkg/sql/logictest/testdata/not_visible_index`.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 95 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zf9gT8CSSo2iPsWpZ:-N8zf9gT8CSSo2iPsWp_:bkakq27) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L95)):*
> ```
> idx_v_visible    NULL
> idx_i_invisible  true
> idx_v_invisible  true
> ```

nit: I don't see the need for this test given that you have the `SHOW CREATE TABLE t1` test above.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 229 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zj6F23iZ7btDqhCQX:-N8zj6F3BdBrsUr_GekM:b-8z6tpm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L229)):*
> ```
> # idx_v_visible is selected if it is visible.
> query T
> EXPLAIN(OPT) SELECT * FROM t1 WHERE v = 2
> ```

I think all these tests that use `EXPLAIN (OPT)` could just use regular `EXPLAIN` and still reveal all the necessary information about which indexes are used. Using regular `EXPLAIN` is the convention is execbuilder tests unless you need to display information that it doesn't show.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 300 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zgXoK6d7GzayKeiqV:-N8zgXoK6d7GzayKeiqW:b-ouendw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L300)):*
> ```
> 
> statement ok
> CREATE INDEX idx_v_invisible ON t1(v) WHERE v > 0
> ```

Is this supposed to be NOT VISIBLE?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 302 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zgaF03nYpGhxtNe7o:-N8zgaF03nYpGhxtNe7p:bucal30) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L302)):*
> ```
> CREATE INDEX idx_v_invisible ON t1(v) WHERE v > 0
> 
> # Partial force index is still in effect, and idx_v_invisible is used when query filter implies predicate.
> ```

I think this test is supposed to check that the partial NOT VISIBLE index is not used.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 572 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zhewV1WMPV7BDjqn9:-N8zhewV1WMPV7BDjqnA:bx5kxrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L572)):*
> ```
> 
> ###########################################################################################
> # Invisible index is still used to check for FK constraint. 
> ```

If you're using GoLand, you can configure it to remove trailing whitespace like this when you save a file: 
![image.png](https://files.reviewableusercontent.io/images/reviewable/1128750/LOA-nhgiRj86F_XylRZ9Oyoc/image.png)

If you're using another IDE, it may have a similar option.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 702 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zi3Gz1uLCd014rk7t:-N8zi3Gz1uLCd014rk7u:b-6432s3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L702)):*
> ```
> 
> query TTB
> SELECT index_name, column_name, visible FROM [SHOW INDEX FROM parent] ORDER BY index_name, seq_in_index
> ```

I don't understand the motivation for these tests with `SHOW INDEX` here and below. What are they proving?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 724 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zjsEBB273ha2YkL5W:-N8zjsEBB273ha2YkL5X:b-1oenj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L724)):*
> ```
> # c_update_idx_invisible is used for constraint check (delete on a parent table requires FK check).
> query T
> EXPLAIN(OPT) DELETE FROM parent where p = 2
> ```

If you use regular `EXPLAIN` without `(OPT)` it'll print information about the cascade which would be useful for these tests.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 858 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zk2Ro2Xxpkey5uuda:-N8zk2Ro2Xxpkey5uudb:bfyzytr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L858)):*
> ```
> CREATE INDEX c_idx_invisible_2 ON child(c) NOT VISIBLE
> ----
> NOTICE: not visible indexes may still be used for unique or foreign key constraint check, so the query plan may be different from dropping the index completely.
> ```

nit: it's odd that the notice mentions dropping and index when one is being added. I wonder if there is a better way to phrase this notice.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Getting closer!

> Reviewed 2 of 56 files at r13, 4 of 6 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-:-N8znr8VD9MGk-RDWSA1:b-b3xry2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, @rhu713, and @wenyihu6)

___
*[`pkg/sql/create_index.go` line 745 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8j9j_p-nc8F8R7lIwG:-N8zfgxbDDMUJUR50Z2Z:b8op14j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6082e6c2e75f04b9b049c5c3205cea10c579e55e/pkg/sql/create_index.go#L745)):*
> But based on the code, it is checking if originColIDs is the prefix of the index's KeyColumnIDs. I'm not sure if that's what we want. Consider the following case: a, b might not be a, c, b's prefix, but it's still useful for constraint check that reference column a, b.

Great example. Yes, looks like `IsValidOriginIndex` will miss some useful indexes. (I suspect it was written back when FK checks were more picky about the indexes they used.)

> If originColIDs is not an ordered list but just in the same order that the user specified, this might cause problems mentioned in the screenshot.

Yes, I think it's just the order the user specified.

> I'm not exactly sure how IsValidOriginIndex and HasPrefix are being used right now; I took a quick look at it, but I need more time to understand better.

Looks like `IsValidOriginIndex` is currently only used in one place, this check in `DROP INDEX`: https://github.com/cockroachdb/cockroach/blob/3aea0c49e466e621237b2ffcbb36381cd7850dca/pkg/sql/drop_index.go#L415-L429

> I'm not sure if this is supposed to happen. When the table gets really large, will it ever favour the the secondary index and perform a look up join?

Sometimes it will. Here's an example: https://gist.github.com/michae2/2bf255b183b529bea28f25aa82b5f561 which results in this for me:

```
                                                 info
------------------------------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • root
  │
  ├── • update
  │   │ table: p
  │   │ set: y
  │   │
  │   └── • buffer
  │       │ label: buffer 1
  │       │
  │       └── • render
  │           │
  │           └── • filter
  │               │ estimated row count: 1
  │               │ filter: y = 3
  │               │
  │               └── • scan
  │                     estimated row count: 100 (100% of the table; stats collected 20 minutes ago)
  │                     table: p@p_pkey
  │                     spans: FULL SCAN
  │
  └── • constraint-check
      │
      └── • error if rows
          │
          └── • lookup join (semi)
              │ estimated row count: 1
              │ table: t@t_pkey
              │ equality: (rowid) = (rowid)
              │ equality cols are key
              │ pred: x = a
              │
              └── • lookup join
                  │ estimated row count: 1
                  │ table: t@t_b_idx
                  │ equality: (y) = (b)
                  │
                  └── • except
                      │ estimated row count: 1
                      │
                      ├── • scan buffer
                      │     estimated row count: 1
                      │     label: buffer 1
                      │
                      └── • scan buffer
                            estimated row count: 1
                            label: buffer 1
```

So I don't think we need `KeyColumnIDs` to cover `originColIDs`. An intersection will work, as you've done. Or checking that the prefix of `KeyColumnIDs` is in `originColIDs` would also work, as @mgartner pointed out.

> Sorry if I am making things really confusing here.

Not confusing at all! This is great! :slightly_smiling_face:
___
*[`pkg/sql/catalog/descpb/index.go` line 73 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zce2SAHplPD_OGAqR:-N8zm8zM3e6KIrGhHw3x:bud84wh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/catalog/descpb/index.go#L73)):*
> checking that a prefix of the indexes columns are included in the FK constraint

Makes sense to me. (Would that be equivalent to just checking if the first column of the index is in the FK constraint?)

Also, I think we should make this change directly to `IsValidOriginIndex` rather than creating a new function. We also want the other check using `IsValidOriginIndex` to pick up this change. (There's only one other check using this, in `DROP INDEX`.)
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 812 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8ys6AaCSqkORtRNkbj:-N8ys6Ab-ZjRT1Vsr8c2:b-jf76fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/catalog/tabledesc/validate.go#L812)):*
> ```Go
> 	noticeStr := "not visible indexes may still be used for unique or " +
> 		"foreign key constraint check, so the query plan may be different from " +
> 		"dropping the index completely."
> ```

nit: This should be `const noticeStr = "not visible ..."`.
___
*[`pkg/sql/opt/optbuilder/select.go` line 533 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zmwnZ2vyqa2pKQK-n:-N8zmwnZ2vyqa2pKQK-o:b-hmxhzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/optbuilder/select.go#L533)):*
> ```Go
> 					//if tab.Index(i).IsNotVisible() {
> 					//	// TODO(wenyihu6): Find a way to give notice here is force index is not visible.
> 					//}
> ```

nit: I think the TODO is enough, no need for the commented-out code.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/alter_table.go` line 865 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8ze8HLBYsBdNcg2LZM:-N8zn1iNC2A-u3DWE1aW:b-192usm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/alter_table.go#L865)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I don't understand this comment and code. Don't we only want to show a notice if an invisible index is being dropped that could be helpful for FK checks and cascades?
</blockquote></details>

I should make this comment better; This check was more for ALTER TABLE ... ADD CONSTRAINT ... FOREIGN KEY ... (If we add a foreign key constraint to a table with invisible indexes, it's giving a notice) I'm giving a notice every time when there is an invisible index created that could be used for constraint check. That might be too noisy; giving a notice every time when the invisible index is dropped is probably better. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/catalog/descpb/index.go` line 73 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zce2SAHplPD_OGAqR:-N8zohX013vm4SMm2_C-:b-cn2a4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/catalog/descpb/index.go#L73)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

> checking that a prefix of the indexes columns are included in the FK constraint

Makes sense to me. (Would that be equivalent to just checking if the first column of the index is in the FK constraint?)

Also, I think we should make this change directly to `IsValidOriginIndex` rather than creating a new function. We also want the other check using `IsValidOriginIndex` to pick up this change. (There's only one other check using this, in `DROP INDEX`.)
</blockquote></details>

That's a good idea. We should also look into where [`HasPrefix`](https://github.com/cockroachdb/cockroach/blob/d4d697a7ed71b672a1c69b4a0cab1db3d1938ae6/pkg/sql/catalog/descpb/structured.go#L122) are used. Their use cases have been really similar to the logic in `IsValidOriginIndex `.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/catalog/descpb/structured.go` line 152 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zcOSLCqo48Vq-oNLw:-N8zpRJABxy3NjKnLKvn:b-a0n7u5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/catalog/descpb/structured.go#L152)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: You can name this `Intersects` to match similar functions like `FastIntSet.Intersects`.
</blockquote></details>

Sounds good. I will make the changes together after our meeting tomorrow.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 572 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zhewV1WMPV7BDjqn9:-N8zqOE2EZOtQxpBeS9j:bru7tu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L572)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

If you're using GoLand, you can configure it to remove trailing whitespace like this when you save a file: 
![image.png](https://files.reviewableusercontent.io/images/reviewable/1128750/LOA-nhgiRj86F_XylRZ9Oyoc/image.png)

If you're using another IDE, it may have a similar option.
</blockquote></details>

Thank you! I just updated my configuration.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 858 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zk2Ro2Xxpkey5uuda:-N8zqtSZCCKrbZxUJlOo:b-y2jg2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L858)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: it's odd that the notice mentions dropping and index when one is being added. I wonder if there is a better way to phrase this notice.
</blockquote></details>

Is it better if we send notices when there are invisible indexes dropped instead?


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 702 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zi3Gz1uLCd014rk7t:-N8zruh979OKK3Wj055z:bidnkh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L702)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I don't understand the motivation for these tests with `SHOW INDEX` here and below. What are they proving?
</blockquote></details>

Ahh I initially used `EXPLAIN`, and there was this `child_delete_p_fkey` shown in the `EXPLAIN` output for cascade check. I was worried that  `child_delete_p_fkey` was some indexes created automatically, so I added those `SHOW INDEX`. I should have removed them.

![Screen Shot 2022-08-08 at 8.00.56 PM.png](https://files.reviewableusercontent.io/images/reviewable/56973754/dzs-89XKAAtGOpJ_HVl7GVne/Screen%20Shot%202022-08-08%20at%208.00.56%20PM.png)

I changed to `EXPLAIN (OPT)` afterwards because that was the only way to see the scan flags.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/create_index.go` line 745 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8j9j_p-nc8F8R7lIwG:-N8zvilo0Nfpbf1e-XaO:b-qm7fjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6082e6c2e75f04b9b049c5c3205cea10c579e55e/pkg/sql/create_index.go#L745)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

> But based on the code, it is checking if originColIDs is the prefix of the index's KeyColumnIDs. I'm not sure if that's what we want. Consider the following case: a, b might not be a, c, b's prefix, but it's still useful for constraint check that reference column a, b.

Great example. Yes, looks like `IsValidOriginIndex` will miss some useful indexes. (I suspect it was written back when FK checks were more picky about the indexes they used.)

> If originColIDs is not an ordered list but just in the same order that the user specified, this might cause problems mentioned in the screenshot.

Yes, I think it's just the order the user specified.

> I'm not exactly sure how IsValidOriginIndex and HasPrefix are being used right now; I took a quick look at it, but I need more time to understand better.

Looks like `IsValidOriginIndex` is currently only used in one place, this check in `DROP INDEX`: https://github.com/cockroachdb/cockroach/blob/3aea0c49e466e621237b2ffcbb36381cd7850dca/pkg/sql/drop_index.go#L415-L429

> I'm not sure if this is supposed to happen. When the table gets really large, will it ever favour the the secondary index and perform a look up join?

Sometimes it will. Here's an example: https://gist.github.com/michae2/2bf255b183b529bea28f25aa82b5f561 which results in this for me:

```
                                                 info
------------------------------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • root
  │
  ├── • update
  │   │ table: p
  │   │ set: y
  │   │
  │   └── • buffer
  │       │ label: buffer 1
  │       │
  │       └── • render
  │           │
  │           └── • filter
  │               │ estimated row count: 1
  │               │ filter: y = 3
  │               │
  │               └── • scan
  │                     estimated row count: 100 (100% of the table; stats collected 20 minutes ago)
  │                     table: p@p_pkey
  │                     spans: FULL SCAN
  │
  └── • constraint-check
      │
      └── • error if rows
          │
          └── • lookup join (semi)
              │ estimated row count: 1
              │ table: t@t_pkey
              │ equality: (rowid) = (rowid)
              │ equality cols are key
              │ pred: x = a
              │
              └── • lookup join
                  │ estimated row count: 1
                  │ table: t@t_b_idx
                  │ equality: (y) = (b)
                  │
                  └── • except
                      │ estimated row count: 1
                      │
                      ├── • scan buffer
                      │     estimated row count: 1
                      │     label: buffer 1
                      │
                      └── • scan buffer
                            estimated row count: 1
                            label: buffer 1
```

So I don't think we need `KeyColumnIDs` to cover `originColIDs`. An intersection will work, as you've done. Or checking that the prefix of `KeyColumnIDs` is in `originColIDs` would also work, as @mgartner pointed out.

> Sorry if I am making things really confusing here.

Not confusing at all! This is great! :slightly_smiling_face:
</blockquote></details>

Thanks a lot for the example! That cleared things up. The rest sounds good to me as well. We can discuss this in more details during our meeting tomorrow.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
> I left a few comments on the notice-related logic that might be helpful, but we should talk about it IRL.
> 
> Great work on this! Nice to see everything coming together!
> 
> All the NOTICE-related changes would be a good candidate to break into a separate commit or PR. It's up to you whether you want to split them up now, but something to keep in mind for future PRs.

Sure! I can split them up; I will keep this PR for notice related changes and open a new PR with changes in the optimizer.

wenyihu6 commented:
I'm not sure if we want this for this release. I'm happy to close this PR, and you can decide if this is needed later on. @mgartner @michae2

michae2 commented:
I'm in favor, WDYT @mgartner?

wenyihu6 commented:
*[`pkg/sql/catalog/descpb/index.go` line 73 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zce2SAHplPD_OGAqR:-NA5Z9_u53LMOP1bYj4T:b-5wmj9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/catalog/descpb/index.go#L73)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

That's a good idea. We should also look into where [`HasPrefix`](https://github.com/cockroachdb/cockroach/blob/d4d697a7ed71b672a1c69b4a0cab1db3d1938ae6/pkg/sql/catalog/descpb/structured.go#L122) are used. Their use cases have been really similar to the logic in `IsValidOriginIndex `.
</blockquote></details>

I ended up leaving `IsValidOriginIndex` just as the way it is because I was worried about breaking the logic now. But we can update this for future release if needed. I will create an issue for this.


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
> I'm in favor of getting this into the release, WDYT @mgartner?

Yes, let's try to get it in.

> Reviewed 1 of 37 files at r8, 16 of 26 files at r9, 1 of 10 files at r10, 1 of 9 files at r12, 46 of 56 files at r13, 2 of 11 files at r16, 12 of 12 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-:-NABa7db37wQTt9IbUEL:boha3up)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rhu713, and @wenyihu6)

___
*[`pkg/sql/catalog/descpb/index.go` line 73 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zce2SAHplPD_OGAqR:-N91txRMAYWXaJaztDIZ:b-mffzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/catalog/descpb/index.go#L73)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

That's a good idea. We should also look into where [`HasPrefix`](https://github.com/cockroachdb/cockroach/blob/d4d697a7ed71b672a1c69b4a0cab1db3d1938ae6/pkg/sql/catalog/descpb/structured.go#L122) are used. Their use cases have been really similar to the logic in `IsValidOriginIndex `.
</blockquote></details>

The error returned in that DROP INDEX case is also confusing to me. I'm questioning 
___
*[`pkg/sql/catalog/descpb/index.go` line 67 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-NABXUdsENZdQ1LiH-gw:-NABXUdt6pouDZlP5-je:bogitov) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2a587c9dda815da19614087780841c140ce060/pkg/sql/catalog/descpb/index.go#L67)):*
> ```Go
> }
> 
> // IsHelpfulOriginIndex returns whether the index may become a helpful index for
> ```

nit: "become" => "be"
___
*[`pkg/sql/catalog/descpb/index.go` line 72 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-NAB_5O_0ZU2VfzwT9KL:-NAB_5O_0ZU2VfzwT9KM:bumqbnn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2a587c9dda815da19614087780841c140ce060/pkg/sql/catalog/descpb/index.go#L72)):*
> ```Go
> // column covered by the index is present in FK constraint columns.
> func (desc *IndexDescriptor) IsHelpfulOriginIndex(originColIDs ColumnIDs) bool {
> 	return !desc.IsPartial() && ColumnIDs(desc.KeyColumnIDs).HasFirstElementIn(originColIDs)
> ```

nit: HasFirstElementIn is so simple I don't think a helper function is necessary. You can replace with `len(desc.KeyColumnIDs) > 0 && originColIDs.Contains(desc.KeyColumnIDs[0])`.
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 780 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-NAB_SB441tbcWiHlt_q:-NAB_SB441tbcWiHlt_r:b-3h2km) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2a587c9dda815da19614087780841c140ce060/pkg/sql/catalog/tabledesc/validate.go#L780)):*
> ```Go
> 	index catalog.Index, tableDesc catalog.TableDescriptor,
> ) pgnotice.Notice {
> 	noticeStr := "not visible indexes may still used to police constraint check"
> ```

nit: police => enforce

Can we have more specific notices for two cases. For example:

1. "some queries may read from a unique, not visible index to enforce uniqueness of the columns"
2. "some queries may read from a not visible index to enforce foreign key constraints"
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 1390 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-NABYUao2a1gLu1dpbzI:-NABYUao2a1gLu1dpbzJ:b1nz3pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2a587c9dda815da19614087780841c140ce060/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L1390)):*
> ```
> 
> query T noticetrace
> CREATE UNIQUE INDEX idx_invisible_unique ON t1(p) VISIBLE
> ```

A `statement ok` for the `CREATE TABLE` and `CREATE UNIQUE INDEX` makes more sense here since you're not testing the notices of them. You can also combine them into the same statement (a single `CREATE TABLE`) or `statement ok` block.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 1403 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-NABZ6t2ElXYmX_Pdunu:-NABZ6t2ElXYmX_Pdunv:bypyaul) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2a587c9dda815da19614087780841c140ce060/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L1403)):*
> ```
> DROP INDEX idx_invisible_unique
> ----
> NOTICE: not visible indexes may still used to police constraint check
> ```

This notice seems out of place - it doesn't make sense in this context. A dropped index cannot be used to enforce a constraint.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 1422 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-NABZOyG4AWk-fwSVgam:-NABZOyG4AWk-fwSVgan:bjxl1gy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2a587c9dda815da19614087780841c140ce060/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L1422)):*
> ```
> DROP TABLE t1
> 
> query T noticetrace
> ```

statement ok
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 1426 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-NABZQLf-V0Vufk-O8de:-NABZQLf-V0Vufk-O8df:bjxl1gy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2a587c9dda815da19614087780841c140ce060/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L1426)):*
> ```
> ----
> 
> query T noticetrace
> ```

statement ok
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 1427 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-NABZaHw80Y7LEY6nCOv:-NABZaHw80Y7LEY6nCOw:bi3zaox) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2a587c9dda815da19614087780841c140ce060/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L1427)):*
> ```
> 
> query T noticetrace
> CREATE TABLE child (c INT PRIMARY KEY, p INT NOT NULL REFERENCES parent(p), INDEX c_idx_invisible_helpful (p, c) NOT VISIBLE, INDEX c_idx_invisible_not_helpful (c, p) NOT VISIBLE)
> ```

Can you break this up into a multi-line statement to make it easier to read?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 1440 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-NABZTUz62QvL4n7Z2wZ:-NABZTUz62QvL4n7Z2w_:b5uxr9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2a587c9dda815da19614087780841c140ce060/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L1440)):*
> ```
> DROP INDEX c_idx_invisible_helpful
> ----
> NOTICE: not visible indexes may still used to police constraint check
> ```

This notice is confusing in this context.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/catalog/tabledesc/validate.go` line 780 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-NAB_SB441tbcWiHlt_q:-NADY7hR2TmBJOlnJHcC:b-oroi6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2a587c9dda815da19614087780841c140ce060/pkg/sql/catalog/tabledesc/validate.go#L780)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: police => enforce

Can we have more specific notices for two cases. For example:

1. "some queries may read from a unique, not visible index to enforce uniqueness of the columns"
2. "some queries may read from a not visible index to enforce foreign key constraints"
</blockquote></details>

These two cases might not be exclusive since a unique index on a parent table might be useful for both uniqueness or foreign key check. The new changes include three types notices:
1. unique index on a parent table that can be helpful for FK check and uniqueness check 
2. other unique indexes that can only be helpful for uniqueness check 
3. other indexes on a child table that can be helpful for FK check

A thing to note --- we are giving out these notices when users are changing an existing index to not visible or dropping an invisible index but not when users execute things like alter table ... add constraint. It is possible that the user adds a foreign key constraint later on. These invisible indexes might not be useful now but may become useful later. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 300 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zgXoK6d7GzayKeiqV:-NAD_oRmBjjP2QvcJBvM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L300)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Is this supposed to be NOT VISIBLE?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 1440 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-NABZTUz62QvL4n7Z2wZ:-NADa2Y6ABvNycFQRlnK:b-r0ft0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2a587c9dda815da19614087780841c140ce060/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L1440)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This notice is confusing in this context.
</blockquote></details>

Are they better now?


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 724 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zjsEBB273ha2YkL5W:-NADaJRS6cL9susrkcXF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L724)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

If you use regular `EXPLAIN` without `(OPT)` it'll print information about the cascade which would be useful for these tests.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 302 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-N8zgaF03nYpGhxtNe7o:-NADaL_Y5jEJ6G9YReZy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8549ad2ee25f66cfc963c805ee11ebf322ed0272/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L302)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I think this test is supposed to check that the partial NOT VISIBLE index is not used.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
I think we should wait for @mgartner before checking this in. My opinions are:
* I find the notices on `DROP` confusing. It seems like at the time of dropping it's already too late for a notice. My preference would be to only have these notices on `ALTER`.
* I think the notices are a bit too verbose. I'd prefer it if we got rid of the "dropping this index may be different from marking this index not visible;" part.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-:-NAR5w5sEpGixAd7vpyN:b-87fj6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rhu713, and @wenyihu6)



<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Thanks for the reply! Yeah. We should wait for Marcus's stamp; he mentioned that he wants to have another look before merging it in our last meeting.





<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Possible notice messages:

First, we need to decide when we want to give notices 
1. ALTER INDEX idx_invisible_unique NOT VISIBLE
2. DROP INDEX idx_invisible_unique
3. Both

Second, we need to decide what notices to give
There are three cases where invisible indexes might be different from dropping the index. 
1. Unique index to enforce uniqueness 
2. Index on child table to enforce FK 
3. Unique index on parent table to enforce uniqueness or FK

We can give out the same message for different cases:
1. queries may still use not visible indexes to enforce constraints check
Or give different messages for different cases:
2. queries may still use not visible indexes to enforce uniqueness of columns
queries may still use not visible indexes to enforce foreign key constraints
queries may still use not visible indexes to enforce unique or foreign key constraints

In addition to the message above, we need to decide if we want to add the following message in front:
1. dropping this index may be different from marking this index not visible
2. Or not 

My preference:
1. Only give notices for ALTER INDEX since if they want to test the behavior of dropping an index, they will need to alter the index visible before dropping it. And this notice is already too late when they are dropping the index. 
2. If we are giving notices only for ALTER INDEX, I prefer "queries may still use not visible indexes to enforce constraints check"
3. If we are giving notices for both ALTER INDEX and DROP INDEX, I prefer 
"dropping this index may be different from marking this index not visible because queries may still use not visible indexes to enforce constraints check"
Giving out different messages for different cases is nice, but getting different messages executing the same SQL statement might lead to confusion. By the time the users are executing `ALTER INDEX …`, they might not remember if that index was unique or on parent or on child table any more. We can write “please read documentation[insert link] for more details” as well. 





<!-- Sent from Reviewable.io -->

mgartner commented:
I agree with you both. My preference is a simple notice for `ALTER INDEX ... NOT VISIBLE` only with a message like "queries may still use not visible indexes to enforce unique and foreign key constraints".

mgartner reviewed (approved):
Looks great! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 5 files at r18.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85354#-:-NAkPyrnBom06K2FGgw7:blerph5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @rhu713, and @wenyihu6)



<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTRs!

bors r+
